packages/open-truss/src/uqi/uqi.ts

```diff
@@ -0,0 +1,176 @@
+import Iterator from './iterator'
+
+export type UqiMappedType =
+  | 'String'
+  | 'Number'
+  | 'Boolean'
+  | 'BigInt'
+  | 'Date'
+  | 'Object'
+
+export type UqiTypeMappings = Record<string, UqiMappedType>
+
+export interface UqiColumn {
+  name: string
+  type: string
+}
+
+export type UqiScalar = string | number | boolean | bigint | object | null
+
+export interface UqiClient {
+  query: (
+    query: string,
+    options?: { statusCallback?: (status: UqiStatus) => Promise<void> },
+  ) => Promise<AsyncIterableIterator<UqiResult>>
+  teardown: () => Promise<void>
+}
+
+interface UqiSettings<C, T> {
+  typeMappings: UqiTypeMappings
+  config: C
+  client: T
+  query: (
+    context: UqiContext<C, T>,
+    query: string,
+  ) => Promise<AsyncIterableIterator<UqiResult>>
+  teardown: (context: UqiContext<C, T>) => Promise<void>
+}
+
+export interface UqiContext<C, T> {
+  config: C
+  client: T
+  status: UqiStatus
+  typeMappings: UqiTypeMappings
+}
+
+export interface UqiResult {
+  row: UqiScalar[]
+  metadata: {
+    columns: UqiColumn[]
+  }
+}
+
+export interface UqiStatus {
+  completedAt: Date | null
+  failedAt: Date | null
+  failedReason: string | null
+  recordsReturned: number
+  startedAt: Date | null
+}
+
+export default function uqi<C, T>(og: UqiSettings<C, T>): UqiClient {
+  const context: UqiContext<C, T> = {
+    config: og.config,
+    client: og.client,
+    status: {
+      completedAt: null,
+      failedAt: null,
+      failedReason: null,
+      recordsReturned: 0,
+      startedAt: null,
+    },
+    typeMappings: og.typeMappings,
+  }
+
+  function buildRow(result: UqiResult): UqiScalar[] {
+    return result.metadata.columns.map((column: UqiColumn, i: number) => {
+      const type = context.typeMappings[column.type]
+      if (!type) {
+        throw new Error(`Type ${column.type} is not mapped`)
+      }
+      const value = result.row[i]
+      if (value === null) {
+        return null
+      }
+      if (typeof value === 'string' && type === 'Number') {
```

Comment:
How do you think this logic will work when we have untyped data sources like rest apis?

Reply:
I don't know but I'm excited to find out! I'm guessing when someone is using a REST API they'll have to define an interface of some sort ahead of time that gets passed in the create function for the uqi rest client. Or maybe we have to bring in zod 🤔

Reply:
Specifying the type of a rest endpoint is what I think 👍